torch/csrc/autograd/python_torch_functions_manual.cpp

```diff
@@ -405,12 +405,27 @@ static PyObject* THPVariable__to_functional_tensor(
     PyObject* kwargs) {
   HANDLE_TH_ERRORS
   static PythonArgParser parser(
-      {"_to_functional_tensor(Tensor t)"}, /*traceable=*/true);
+      {"_to_functional_tensor(Tensor t, *, bool mirror_autograd_meta=False)"},
+      /*traceable=*/true);
 
-  ParsedArgs<1> parsed_args;
+  ParsedArgs<2> parsed_args;
   auto r = parser.parse(args, kwargs, parsed_args);
   auto self_ = r.tensor(0);
+  auto mirror_autograd_meta = r.toBool(1);
   auto wrapped = at::functionalization::impl::to_functional_tensor(self_);
+  if (mirror_autograd_meta) {
+    // Here, we unsafely set the grad function on the wrapper to be the same as
+    // the inner. We expect this grad_fn to NEVER be used. It's needed so that
+    // .is_leaf metadata is accurate on the wrapper
+    auto inner_autograd_meta = impl::get_autograd_meta(self_);
+    if (inner_autograd_meta) {
+      wrapped.set_requires_grad(self_.requires_grad());
+      if (wrapped.requires_grad()) {
+        impl::get_autograd_meta(wrapped)->grad_fn_ =
+            inner_autograd_meta->grad_fn_;
```

Comment:
TODO: this is dangerous, and I'm going to change it to an "error on invocation grad fn". I think that exists somewhere in the code base, and just need to find it.

The idea is that if we have inputs to aot autograd that require gradients, that might have their own grad fn, then they will be marked with `input.is_leaf == False`, and we want the same to be true for our FunctionalWrapper inputs that we trace with.

However, the grad fn that we put on the input tensor wrappers should never be accessed. When we trace through the backward to create our joint graph, we specify the graph inputs as the primals, so autograd should stop there. Making the grad fn here an "error function" should assert that.